File: python/String-2/count_hi.py
# Return the number of times that the string "hi" appears anywhere in the given string.

# def count_hi(str):
#   return str.count('hi')

def count_hi(str):
  return count(str, 'hi')

def count(str, term):
  count = 0
  for i in range(len(str)-len(term)+1):
    if str[i:i+len(term)] == term:
      count +=1
  return count

def test_func(func, s, expected):
  result = func(s)
  return "{0}({1:>18}) → {2:2} == {3:2} ? {4:2}".format(func.__name__, s, expected, result, "OK" if result == expected else "X")

def test_cases():
  return [
    ['abc hi ho', 1],
    ['ABChi hi', 2],
    ['hihi', 2],
    ['hiHIhi', 2],
    ['', 0],
    ['h', 0],
    ['hi', 1],
    ['Hi is no HI on ahI', 0],
    ['hiho not HOHIhi', 2]
  ]

def test():
  for t in test_cases():
    print(test_func(count_hi, t[0], t[1]))

if __name__ == "__main__":
  test()

File: java/src/main/com/codingbat/string2/CountHi.java
package main.com.codingbat.string2;

import java.util.HashMap;
import java.util.Map;

public class CountHi {
  /**
   * Return the number of times that the string "hi" appears anywhere in the
   *  given string.
   * 
   * @param str String of characters.
   * @return Count of "hi" substrings.
   */
  public static int countHi(String str) {
    Map<String, Integer> map = new HashMap<String, Integer>();
    for (String s : new String[] { "hi" }) {
      map.put(s, 0);
      for (int i=0; i < str.length() - s.length()+1; i++) {
        if (str.substring(i, i+s.length()).equals(s)) {
          map.put(s, map.get(s) + 1);
        }
      }
    }
    return map.get("hi");
  }
  
}
